File: backend/app/domains/metadata/schemas/metadata.py
"""Pydantic schemas for taxonomy/metadata management (F256)."""

from __future__ import annotations

import json
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator, model_validator

MetadataFieldType = Literal["text", "select", "multi_select", "date", "boolean", "number"]


class CreateMetadataFieldRequest(BaseModel):
    name: str = Field(min_length=1, max_length=128, pattern=r"^[a-z0-9_]+$")
    display_name: str = Field(min_length=1, max_length=256)
    field_type: MetadataFieldType
    allowed_values: list[str] | None = None
    is_required: bool = False
    is_filterable: bool = True
    description: str | None = Field(default=None, max_length=1024)
    sort_order: int = Field(default=0, ge=0)

    @field_validator("display_name")
    @classmethod
    def strip_display_name(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("display_name must not be blank")
        return stripped

    @model_validator(mode="after")
    def validate_allowed_values(self) -> "CreateMetadataFieldRequest":
        if self.field_type in ("select", "multi_select"):
            if not self.allowed_values:
                raise ValueError("allowed_values is required for select/multi_select fields")
            if len(self.allowed_values) > 200:
                raise ValueError("allowed_values must contain at most 200 entries")
        else:
            if self.allowed_values is not None:
                raise ValueError(f"allowed_values is not valid for field_type '{self.field_type}'")
        return self


class UpdateMetadataFieldRequest(BaseModel):
    display_name: str | None = Field(default=None, min_length=1, max_length=256)
    allowed_values: list[str] | None = None
    is_required: bool | None = None
    is_filterable: bool | None = None
    description: str | None = Field(default=None, max_length=1024)
    sort_order: int | None = Field(default=None, ge=0)
    is_active: bool | None = None

    @field_validator("display_name")
    @classmethod
    def strip_display_name(cls, v: str | None) -> str | None:
        if v is None:
            return v
        stripped = v.strip()
        if not stripped:
            raise ValueError("display_name must not be blank")
        return stripped


class MetadataFieldResponse(BaseModel):
    field_id: str
    organization_id: str
    name: str
    display_name: str
    field_type: MetadataFieldType
    allowed_values: list[str] | None
    is_required: bool
    is_filterable: bool
    description: str | None
    sort_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime


class MetadataFieldListResponse(BaseModel):
    items: list[MetadataFieldResponse]
    total: int


# ── Document metadata values ───────────────────────────────────────────────────


class MetadataValueIn(BaseModel):
    """A single field-value pair in a set-metadata request."""

    field_id: str
    value: str | list[str] | bool | float | None

    @field_validator("field_id")
    @classmethod
    def validate_field_id(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("field_id must not be blank")
        return v.strip()


class SetDocumentMetadataRequest(BaseModel):
    """Replaces (upserts) metadata values for a single document."""

    values: list[MetadataValueIn] = Field(min_length=1)


class BulkSetMetadataRequest(BaseModel):
    """Set the same metadata values on multiple documents at once."""

    document_ids: list[str] = Field(min_length=1, max_length=500)
    values: list[MetadataValueIn] = Field(min_length=1)

    @field_validator("document_ids")
    @classmethod
    def validate_document_ids(cls, v: list[str]) -> list[str]:
        cleaned = [x.strip() for x in v if x.strip()]
        if not cleaned:
            raise ValueError("document_ids must not be empty")
        return cleaned


class DocumentMetadataValueResponse(BaseModel):
    field_id: str
    field_name: str
    display_name: str
    field_type: MetadataFieldType
    value: str | list[str] | bool | float | None
    updated_at: datetime


class DocumentMetadataResponse(BaseModel):
    document_id: str
    values: list[DocumentMetadataValueResponse]


class BulkSetMetadataResponse(BaseModel):
    updated: int
    skipped: int
    errors: list[str]


# ── Audit log ─────────────────────────────────────────────────────────────────


class MetadataAuditEntryResponse(BaseModel):
    audit_id: str
    document_id: str
    field_id: str
    field_name: str
    changed_by_id: str | None
    old_value: str | None
    new_value: str | None
    action: Literal["set", "delete", "bulk_set"]
    created_at: datetime


class MetadataAuditListResponse(BaseModel):
    items: list[MetadataAuditEntryResponse]
    total: int


# ── Tag suggestion helper ──────────────────────────────────────────────────────


class TagSuggestionResponse(BaseModel):
    """Returns candidate values for a select/multi_select field based on a prefix."""

    field_id: str
    prefix: str
    suggestions: list[str]

File: frontend/src/lib/api/metadata.ts
import { apiRequest } from "@/lib/api/request";

export type MetadataFieldType =
  | "text"
  | "select"
  | "multi_select"
  | "date"
  | "boolean"
  | "number";

export type MetadataFieldResponse = {
  field_id: string;
  organization_id: string;
  name: string;
  display_name: string;
  field_type: MetadataFieldType;
  allowed_values: string[] | null;
  is_required: boolean;
  is_filterable: boolean;
  description: string | null;
  sort_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type MetadataFieldListResponse = {
  items: MetadataFieldResponse[];
  total: number;
};

export type CreateMetadataFieldRequest = {
  name: string;
  display_name: string;
  field_type: MetadataFieldType;
  allowed_values?: string[] | null;
  is_required?: boolean;
  is_filterable?: boolean;
  description?: string | null;
  sort_order?: number;
};

export type UpdateMetadataFieldRequest = {
  display_name?: string;
  allowed_values?: string[] | null;
  is_required?: boolean;
  is_filterable?: boolean;
  description?: string | null;
  sort_order?: number;
  is_active?: boolean;
};

export type DocumentMetadataValueResponse = {
  field_id: string;
  field_name: string;
  display_name: string;
  field_type: MetadataFieldType;
  value: string | string[] | boolean | number | null;
  updated_at: string;
};

export type DocumentMetadataResponse = {
  document_id: string;
  values: DocumentMetadataValueResponse[];
};

export type MetadataValueIn = {
  field_id: string;
  value: string | string[] | boolean | number | null;
};

export type SetDocumentMetadataRequest = {
  values: MetadataValueIn[];
};

export type BulkSetMetadataRequest = {
  document_ids: string[];
  values: MetadataValueIn[];
};

export type BulkSetMetadataResponse = {
  updated: number;
  skipped: number;
  errors: string[];
};

export type TagSuggestionResponse = {
  field_id: string;
  prefix: string;
  suggestions: string[];
};

export type MetadataAuditEntryResponse = {
  audit_id: string;
  document_id: string;
  field_id: string;
  field_name: string;
  changed_by_id: string | null;
  old_value: string | null;
  new_value: string | null;
  action: "set" | "delete" | "bulk_set";
  created_at: string;
};

export type MetadataAuditListResponse = {
  items: MetadataAuditEntryResponse[];
  total: number;
};

// ── Admin: taxonomy field CRUD ─────────────────────────────────────────────────

export async function listMetadataFields(
  includeInactive = false,
): Promise<MetadataFieldListResponse> {
  return apiRequest<MetadataFieldListResponse>("/admin/metadata/fields", {
    query: { include_inactive: includeInactive },
  });
}

export async function createMetadataField(
  payload: CreateMetadataFieldRequest,
): Promise<MetadataFieldResponse> {
  return apiRequest<MetadataFieldResponse>("/admin/metadata/fields", {
    method: "POST",
    json: payload,
  });
}

export async function getMetadataField(
  fieldId: string,
): Promise<MetadataFieldResponse> {
  return apiRequest<MetadataFieldResponse>(
    `/admin/metadata/fields/${encodeURIComponent(fieldId)}`,
  );
}

export async function updateMetadataField(
  fieldId: string,
  payload: UpdateMetadataFieldRequest,
): Promise<MetadataFieldResponse> {
  return apiRequest<MetadataFieldResponse>(
    `/admin/metadata/fields/${encodeURIComponent(fieldId)}`,
    { method: "PATCH", json: payload },
  );
}

export async function deleteMetadataField(fieldId: string): Promise<void> {
  await apiRequest<unknown>(
    `/admin/metadata/fields/${encodeURIComponent(fieldId)}`,
    { method: "DELETE" },
  );
}

export async function suggestTagValues(
  fieldId: string,
  prefix: string,
): Promise<TagSuggestionResponse> {
  return apiRequest<TagSuggestionResponse>(
    `/admin/metadata/fields/${encodeURIComponent(fieldId)}/suggest`,
    { query: { prefix } },
  );
}

// ── Document metadata ──────────────────────────────────────────────────────────

export async function getDocumentMetadata(
  documentId: string,
): Promise<DocumentMetadataResponse> {
  return apiRequest<DocumentMetadataResponse>(
    `/documents/${encodeURIComponent(documentId)}/metadata`,
  );
}

export async function setDocumentMetadata(
  documentId: string,
  payload: SetDocumentMetadataRequest,
): Promise<DocumentMetadataResponse> {
  return apiRequest<DocumentMetadataResponse>(
    `/documents/${encodeURIComponent(documentId)}/metadata`,
    { method: "PUT", json: payload },
  );
}

export async function bulkSetMetadata(
  payload: BulkSetMetadataRequest,
): Promise<BulkSetMetadataResponse> {
  return apiRequest<BulkSetMetadataResponse>("/admin/metadata/bulk-set", {
    method: "POST",
    json: payload,
  });
}

export type FilterDocumentsResponse = {
  document_ids: string[];
};

export async function filterDocumentsByMetadata(
  filters: Array<{ fieldId: string; value: string }>,
): Promise<FilterDocumentsResponse> {
  const params = filters.map((f) => `${f.fieldId}:${f.value}`);
  return apiRequest<FilterDocumentsResponse>(
    "/admin/metadata/filter-documents",
    {
      query: { filter: params },
    },
  );
}

export async function getDocumentMetadataAudit(
  documentId: string,
  options: { limit?: number; offset?: number } = {},
): Promise<MetadataAuditListResponse> {
  return apiRequest<MetadataAuditListResponse>(
    `/documents/${encodeURIComponent(documentId)}/metadata/audit`,
    { query: { limit: options.limit, offset: options.offset } },
  );
}
